Add kubelet stats receiver

pmcollins opened the pull request:
Please see README for details.

Commit:
Add kubelet stats receiver

Please see README for details.

jrcamp commented on receiver/kubeletstatsreceiver/kubelet/client.go:
```diff
@@ -0,0 +1,205 @@
+	AuthType AuthType `mapstructure:"auth_type"`
+	// CacertPath should be the path to the root CAs. Used to verify the kubelet
+	// server's certificates.
+	CacertPath string `mapstructure:"ca_cert_path"`
```
Can you add a TODO to replace this with https://github.com/open-telemetry/opentelemetry-collector/issues/933 once done?

jrcamp commented on receiver/kubeletstatsreceiver/kubelet/client.go:
```diff
@@ -0,0 +1,205 @@
+	}
+	return &Client{
+		c: &http.Client{
+			Transport: &http.Transport{
```
Should clone default?

pmcollins replied:
Ah yes, I updated the first client but not this one. Will fix.

jrcamp commented on receiver/kubeletstatsreceiver/kubelet/client.go:
```diff
@@ -0,0 +1,205 @@
+	}
+	if baseURL == "" {
+		// this requires HostNetwork to be turned on
+		hostname, err := os.Hostname()
```
I think it might better for endpoint to be configured explicitly by the end user or via a discovery rule.

pmcollins replied:
Sounds good to me. This was taken from SmartAgent BTW.

jrcamp commented on receiver/kubeletstatsreceiver/kubelet/network.go:
```diff
@@ -0,0 +1,40 @@
+	}, s.Time.Time)
+}
+
+func rxBytesMetric(s *stats.NetworkStats) *metricspb.Metric {
+	metric := cumulativeInt("network/received", "By", s.RxBytes)
+	applyLabels(metric, map[string]string{"interface": s.Name})
+	return metric
+}
+
+func txBytesMetric(s *stats.NetworkStats) *metricspb.Metric {
+	metric := cumulativeInt("network/transmitted", "By", s.TxBytes)
+	applyLabels(metric, map[string]string{"interface": s.Name})
+	return metric
+}
```
Would align these with https://github.com/open-telemetry/opentelemetry-collector/blob/master/receiver/hostmetricsreceiver/internal/scraper/networkscraper/network_constants.go which has `host/network/bytes` (so maybe `kubernetes/pod/network/bytes`?) with `direction` label of receiver or transmit.

pmcollins replied:
Good to know, thanks. Will take a look.

pmcollins replied:
Done.

jrcamp commented on receiver/kubeletstatsreceiver/kubelet/fs.go:
```diff
@@ -0,0 +1,35 @@
+}
+
+func fsAvailableMetric(s *stats.FsStats) *metricspb.Metric {
+	return intGauge("fs/available", "By", s.AvailableBytes)
```
I wonder if all these be prefixed by `kubernetes/(pod|container)` similar to kubernetesclusterreceiver?

pmcollins replied:
I've prefixed all metrics with `kubernetes/(pod|container|node)`. Or would `k8s` be better than `kubernetes`?

jrcamp replied:
Yeah good point since in semantic conventions they use k8s. @asuresh4 what do you think, should the metrics in clusterreceiver be k8s/*?

asuresh4 replied:
Yeah, I'm good with updating those metric names to start with k8s/. Since that is consistent with the conventions and also more concise.

pmcollins replied:
Done.

pmcollins commented on receiver/kubeletstatsreceiver/factory.go:
```diff
@@ -0,0 +1,68 @@
+func (f Factory) CreateDefaultConfig() configmodels.Receiver {
+	return &Config{
+		ReceiverSettings: configmodels.ReceiverSettings{
+			TypeVal: typeStr, // TODO who tf uses this?
```
Ah, thanks. (this was meant be a comment for myself which I forgot to remove)

pmcollins commented on receiver/kubeletstatsreceiver/kubelet/client.go:
```diff
@@ -0,0 +1,205 @@
+type ClientConfig struct {
+	// AuthType should be "tls" or "serviceAccount". TLS requires that the cert/key
+	// paths below also be set.
+	AuthType AuthType `mapstructure:"auth_type"`
```
Ok, will take a look.

pmcollins commented on receiver/kubeletstatsreceiver/README.md:
```diff
@@ -0,0 +1,65 @@
+    ca_cert_path: "/path/to/ca.crt"
+    client_key_path: "/path/to/apiserver.key"
+    client_cert_path: "/path/to/apiserver.crt"
+    base_url: "https://192.168.64.1:10250"
```
I did end up using endpoint in the code; will update the readme.

jrcamp replied:
Looks like the code still calls it baseURL in places, could it just always refer to it as endpoint to avoid introducing another term?

pmcollins replied:
Ok, I got rid of baseURL everywhere :-)

pmcollins committed:
Use .io import paths, prefix metrics with resource

pmcollins committed:
Add network direction metric labels

pmcollins committed:
Change kubernetes metric prefix to k8s

tigrannajaryan commented on receiver/kubeletstatsreceiver/README.md:
```diff
@@ -0,0 +1,65 @@
+  kubeletstats:
+    collection_interval: 20s
+    auth_type: "tls"
+    ca_cert_path: "/path/to/ca.crt"
+    client_key_path: "/path/to/apiserver.key"
+    client_cert_path: "/path/to/apiserver.crt"
```
There is an upcoming change that uses different key names for client configs: https://github.com/open-telemetry/opentelemetry-collector/pull/988/files/405db7ad13c8feeaeb65142081b04b48e407eaf6#diff-b99ea1d4f9c67b0e6a0dc0199de54d66
It may be useful for you to sync with @ccaraman on this. For now you can probably just use the same field names to avoid breaking changes later once @ccaraman change is merged.

pmcollins replied:
Using the same field names for now. Thanks.

tigrannajaryan commented on receiver/kubeletstatsreceiver/factory.go:
```diff
@@ -0,0 +1,68 @@
+	cfg configmodels.Receiver,
+	nextConsumer consumer.TraceConsumerOld,
+) (component.TraceReceiver, error) {
+	return nil, nil
```
This should return `configerror.ErrDataTypeIsNotSupported`.

pmcollins replied:
Fixed.

pmcollins commented on go.sum:
```diff
@@ -922,6 +923,8 @@ github.com/onsi/gomega v1.7.1/go.mod h1:XdKZgCCFLUoM/7CFJVPcG8C1xQ1AJ0vpAezJrB7J
 github.com/onsi/gomega v1.8.1/go.mod h1:Ho0h+IUsWyvy1OpqCwxlQ/21gkhVunqlU8fDGcoTdcA=
 github.com/onsi/gomega v1.9.0 h1:R1uwffexN6Pr340GtYRIdZmAiN4J+iw6WG4wog1DUXg=
 github.com/onsi/gomega v1.9.0/go.mod h1:Ho0h+IUsWyvy1OpqCwxlQ/21gkhVunqlU8fDGcoTdcA=
+github.com/open-telemetry/opentelemetry-collector v0.3.1-0.20200424155504-9d16f5971ef9 h1:maf2tPAv8QV17gSAePuzREqKvConQ+ozOkNM0pFgbyA=
```
Done, thanks.

pmcollins committed:
Respond to PR comments

Update tls config field names, update go.mod, return appropriate error for
create trace receiver.

pmcollins committed:
Add config validation test

tigrannajaryan commented:
@pmcollins please fix the coverage checks.

pmcollins committed:
Rework kubelet client for testability, add tests

Code coverage was failing the previous version because of lack of tests
of client code.

Previous client implementation was copied from smart agent with several
modifications and was not as easy to test since it did network and file
I/O. This change makes I/O code replaceable making Client easier to
test.

Additionally, ServiceAccount auth has been removed because I wasn't able
to end-to-end test it due to Minikube config issues. It will be added
in a separate PR.

tigrannajaryan commented:
@pmcollins please fix the build.

pmcollins committed:
Improve test coverage
